Feat(eos_designs): Add support to use regex for network services `bgp_peers.nodes`

laxmikantchintakindi opened the pull request:
## Change Summary

Add support for to use regex for network services `bgp_peers.nodes`

## Related Issue(s)

Fixes #3973

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes
When defining nodes for bgp peer groups and bgp peers within my tenant there needs to be the ability to use regex to match all l3leafs instead of defining each one.

## How to test
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->

## Checklist

### User Checklist

<!-- Add your own checklist using MD syntax and by replacing N/A -->
- N/A

### Repository Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.arista.com/stable/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

Commit:
Feat(eos_designs): Add support for to use regex for network services bgp_peers.nodes

laxmikantchintakindi committed:
Merge branch 'devel' into feat/bgp_peer_regex

Commit:
Update tests

Commit:
Add a comment

Commit:
Fix tests

Vibhu-gslab reviewed (commented):
Need for  bgp peer groups too, as asked in the issue.

laxmikantchintakindi commented:
> Need for bgp peer groups too, as asked in the issue.

peer groups are automatically added to nodes when it is used for any peer.  I think it is not required for peer groups.
@ClausHolbechArista @gmuloc

Commit:
Update description

laxmikantchintakindi committed:
Merge branch 'devel' into feat/bgp_peer_regex

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

gmuloc committed:
Merge branch 'devel' into feat/bgp_peer_regex

laxmikantchintakindi committed:
Merge branch 'devel' into feat/bgp_peer_regex

gmuloc commented:
> > Need for bgp peer groups too, as asked in the issue.
> 
> peer groups are automatically added to nodes when it is used for any peer. I think it is not required for peer groups. @ClausHolbechArista @gmuloc

@laxmikantchintakindi / @Vibhu-gslab - the `nodes` key for peer groups is (supposedly) used to restrict on which nodes the peer groups are rendered, otherwise the default behavior is to render only if a peer exist.

Actually the code is different as it is not really a restriction but an `or`:

```python
        peer_peergroups = set()
        for tenant in self.shared_utils.filtered_tenants:
            for vrf in tenant.vrfs:
                # bgp_peers is already filtered in filtered_tenants to only contain entries with our hostname
                if not (vrf.bgp_peers or vrf.bgp_peer_groups):
                    continue

                vrf_peer_peergroups = {peer.peer_group for peer in vrf.bgp_peers if peer.peer_group}
                peer_groups.extend(
                    [
                        peer_group
                        for peer_group in vrf.bgp_peer_groups
                        if self.shared_utils.hostname in peer_group.nodes or peer_group.name in vrf_peer_peergroups
                    ], <<<<<<<<<<<<<<<<<<<<<< notice the `or` here
                )
                peer_peergroups.update(vrf_peer_peergroups)

            peer_groups.extend(
                [peer_group for peer_group in tenant.bgp_peer_groups if self.shared_utils.hostname in peer_group.nodes or peer_group.name in peer_peergroups],
            )  <<<<<<<<<<<<<<<<<<<<<< notice the `or` here and here
```

I feel that having a regex to check the nodes on which to add the peer-group when there is not peer defined is a bit overkill/over-engineered. 
That being said, the issue asked for both and to keep people sane I would implement it for the sake of "it behaves consistently". 

@laxmikantchintakindi can you please:
1. Verify the behavior and correct the schema description as reading the code it seems `nodes`  is not use to restrict anything under peer_groups but rather to generate peer group config when there is not peer (I think the behavior is ok)
2. Add the regex mapping to the nodes under peer group (in tenants and vrfs) so that "it works consistently".

ClausHolbechArista commented:
@gmuloc the `nodes` on peer_groups is inteded as an `or` to **force** a peer-group on a device _even_ when not in use.

gmuloc commented:
> @gmuloc the `nodes` on peer_groups is inteded as an `or` to **force** a peer-group on a device _even_ when not in use.

@laxmikantchintakindi please adjust the descriptions! :)

Commit:
Update description and regex to bgp peer groups

laxmikantchintakindi commented:
> ```python
> <<<<<<<<<<<<<<<<<<<<< notice the `or` here
>                 )
> ```

Added regex support for bgp_peer_groups and updated the description.

Vibhu-gslab commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -114,8 +114,8 @@ $defs:
                 description: |
-                  Nodes is required to restrict configuration of BGP neighbors to certain nodes in the network.
-                  If not set the peer-group is created on devices which have a bgp_peer mapped to the corresponding peer_group.
+                  List of node names or fully matching regex to configure the peer group.
```
Please add an example of regex

laxmikantchintakindi replied:
https://github.com/aristanetworks/avd/pull/5747/files#diff-315c3a77a1954fbd4c36a8488f39397955c4dba40a56382baa12ad7e97f9c8c2R242

Its there in inputs.

laxmikantchintakindi replied:
Done.

Vibhu-gslab reviewed (approved):
Minor comment rest LGTM!

Vibhu-gslab committed:
Merge branch 'devel' into feat/bgp_peer_regex

Commit:
Use match_nodes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Improve coverage

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Fix

Commit:
Revert match nodes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Add regex to tenant level bgp_peer_groups

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Shivani-gslab commented on python-avd/pyavd/_eos_designs/shared_utils/filtered_tenants.py:
```diff
@@ -196,7 +196,7 @@ def filtered_vrfs(
                 continue
 
-            vrf.bgp_peers = vrf.bgp_peers._filtered(lambda bgp_peer: self.hostname in bgp_peer.nodes)._natural_sorted(sort_key="ip_address")
+            vrf.bgp_peers = vrf.bgp_peers._filtered(lambda bgp_peer: self.match_regexes(bgp_peer.nodes, self.hostname))._natural_sorted(sort_key="ip_address")
```
can we use the `match_nodes` util method here, which you had implemented in aggregate_addresses PR

laxmikantchintakindi replied:
Done.

laxmikantchintakindi replied:
This is creating bug. Reverting the change.

laxmikantchintakindi replied:
As we are not creating bgp_peer on all nodes when bgp_peer.nodes is not defined.

laxmikantchintakindi committed:
Merge branch 'devel' into feat/bgp_peer_regex

Commit:
Use match nodes

Shivani-gslab commented on python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py:
```diff
@@ -89,13 +89,17 @@ def _router_bgp_peer_groups(self: AvdStructuredConfigNetworkServicesProtocol) ->
                         peer_group
                         for peer_group in vrf.bgp_peer_groups
-                        if self.shared_utils.hostname in peer_group.nodes or peer_group.name in vrf_peer_peergroups
+                        if self.shared_utils.match_regexes(peer_group.nodes, self.shared_utils.hostname) or peer_group.name in vrf_peer_peergroups
```
same in this file, `match_nodes`

laxmikantchintakindi replied:
This does not work with match_nodes, as match_nodes enforces the bgp_peer_groups when nodes are not defined.

Commit:
revert match nodes

gmuloc committed:
Merge branch 'devel' into feat/bgp_peer_regex

gmuloc committed:
Merge branch 'devel' into feat/bgp_peer_regex

gmuloc commented on ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS/Tenant_C.yml:
```diff
@@ -84,7 +88,7 @@ tenant_c:
               rcf_in: TEST_RCF_IN()
               rcf_out: TEST_RCF_OUT()
-            nodes: [ DC1-BL2B ]
+            nodes: ['DC1-BL[1|2]B']
```
what are you matching with this regex?

laxmikantchintakindi replied:
DC1-BL1B or DC1-BL2B

laxmikantchintakindi replied:
Updated the nodes input with DC1-BL.B

gmuloc replied:
ah :) just commented back :)

Commit:
Update test

gmuloc commented on ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS/Tenant_C.yml:
```diff
@@ -84,7 +88,7 @@ tenant_c:
               rcf_in: TEST_RCF_IN()
               rcf_out: TEST_RCF_OUT()
-            nodes: [ DC1-BL2B ]
+            nodes: ['DC1-BL[1|2]B']
```
I guess the quote is necessary around the regex to prevent the inner `]`  to be interpreted maybe we should mention it somewhere

Also I think your regex will match `DC1-BL|B` which you don't want ;) 


```suggestion
            nodes: ['DC1-BL[12]B']
```

laxmikantchintakindi commented on ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS/Tenant_C.yml:
```diff
@@ -84,7 +88,7 @@ tenant_c:
               rcf_in: TEST_RCF_IN()
               rcf_out: TEST_RCF_OUT()
-            nodes: [ DC1-BL2B ]
+            nodes: ['DC1-BL.B']
```
```suggestion
            nodes: ['DC1-BL[12]B']
```

laxmikantchintakindi committed:
Update ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS/Tenant_C.yml

gmuloc committed:
Merge branch 'devel' into feat/bgp_peer_regex

gmuloc committed:
Merge branch 'devel' into feat/bgp_peer_regex

gmuloc commented on ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS/Tenant_C.yml:
```diff
@@ -84,7 +88,7 @@ tenant_c:
               rcf_in: TEST_RCF_IN()
               rcf_out: TEST_RCF_OUT()
-            nodes: [ DC1-BL2B ]
+            nodes: ['DC1-BL[12]B']
```
should we not see changes to DC1-BL2B as well (need to check other input variables for this one)

laxmikantchintakindi replied:
DC1-BL2B already has this config.

gmuloc replied:
duh .... thanks!

gmuloc commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -114,8 +114,8 @@ $defs:
                 description: |
-                  Nodes is required to restrict configuration of BGP neighbors to certain nodes in the network.
-                  If not set the peer-group is created on devices which have a bgp_peer mapped to the corresponding peer_group.
+                  List of node names or fully matching regex to configure the peer group, e.g. 'DC1-BL[12]B'.
```
```suggestion
                  List of node names or fully matching regexes to configure the peer group, e.g. 'DC1-BL[12]B'.
```

Can we add a test with 2 regexes?

laxmikantchintakindi replied:
Added.

gmuloc commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -1375,7 +1375,9 @@ $defs:
                         type: str
-                      description: Nodes is required to restrict configuration of BGP neighbors to certain nodes in the network.
+                      description: |-
+                        List of node names or fully matching regexes, e.g DC1-BL[12]A.
```
don't you need quotes?

laxmikantchintakindi replied:
Added.

Commit:
Fix descriptions and improve test with multiple regexes.

gmuloc committed:
Merge branch 'devel' into feat/bgp_peer_regex

gmuloc commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -1376,7 +1377,7 @@ $defs:
                         type: str
                       description: |-
-                        List of node names or fully matching regexes, e.g DC1-BL[12]A.
+                        List of node names or fully matching regexes, e.g `DC1-BL[12]A`. The regexes should be written in quotes.
```
```suggestion
                        List of node names or fully matching regexes, e.g `DC1-BL[12]A`.
                        The regexes should be written in quotes.
```

gmuloc committed:
Update python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

gmuloc commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -114,8 +114,9 @@ $defs:
                 description: |
-                  Nodes is required to restrict configuration of BGP neighbors to certain nodes in the network.
-                  If not set the peer-group is created on devices which have a bgp_peer mapped to the corresponding peer_group.
+                  List of node names or fully matching regexes to configure the peer group, e.g. 'DC1-BL[12]B'.
+                  The regexes should be written in quotes.
```
```suggestion
                  List of node names or regular expressions to match against hostnames to configure the peer-group.
                  The regexes need to match full hostname and be enclosed in quotes, e.g 'DC1-BL[12]A'.
```

gmuloc commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -1375,7 +1376,10 @@ $defs:
                         type: str
-                      description: Nodes is required to restrict configuration of BGP neighbors to certain nodes in the network.
+                      description: |-
+                        List of node names or fully matching regexes, e.g `DC1-BL[12]A`.
+                        The regexes should be written in quotes.
```
```suggestion
                  List of node names or regular expressions to match against hostnames.
                  The regexes need to match full hostname and be enclosed in quotes, e.g 'DC1-BL[12]A'.
```

gmuloc commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -1477,8 +1481,9 @@ $defs:
                       description: |
-                        Nodes is required to restrict configuration of BGP neighbors to certain nodes in the network.
-                        If not set the peer-group is created on devices which have a bgp_peer mapped to the corresponding peer_group.
+                        List of node names or fully matching regex to configure the peer group, e.g. 'DC1-BL[12]B'.
+                        The regexes should be written in quotes.
```
```suggestion
                        List of node names or regular expressions to match against hostnames to configure the peer-group.
                        The regexes need to match full hostname and be enclosed in quotes, e.g 'DC1-BL[12]A'.
```

gmuloc committed:
Update python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml

gmuloc committed:
Update python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml

gmuloc committed:
Update python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml

gmuloc committed:
Fix: Schema messed up by myself

laxmikantchintakindi committed:
Merge branch 'devel' into feat/bgp_peer_regex
